### PR TITLE
stdio/format: Fix printf %{g, s, c}. Add %{G, f, F, e, E, a, A}

### PR DESCRIPTION
**Pull request by astalke**
- Remove invalid code that prints float
- Extract floating point printing to a new function
- Add support for conversion specifiers: %f, %F, %e, %E, %G, %a, %A
- Add support for flag: '-'
- Fix: Sign flag now works for positive values.
- Fix: double argument representing infinity is printed as 'inf' for %f, %e and %g conversion specifiers and as 'INF' for %F, %E and %G
- Fix: double argument representing NaN is printed as 'nan' for %f, %e and %g conversion specifiers and as 'NAN' for %F, %E and %G
- Fix: %g conversion specifier now doesn't ignore field width for {0.0, -0.0, NaN, infinity}
- Fix: conversion specifier %c used to ignore minimum field width
- Fix: conversion specifier %s used to truncate strings to minimum field width, instead of precision

DONE: RTOS-318

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by damianloew — COMMENTED** [review]
I am very happy about this PR, but if it's not a big problem handling also %a and %A will be helpful.

**Comment by astalke** [comment]
> I am very happy about this PR, but if it's not a big problem handling also %a and %A will be helpful.

I've just added support for %a and %A

File: stdio/bignum.h
Comment (agkaminski) [inline stdio/bignum.h:35]: Style: Keep two empty lines separation
```suggestion
extern int bignum_init(bignum_t *bn, size_t size, bignum_data_t val);


extern void bignum_destroy(bignum_t *bn);


```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:136]: There's nothing to destroy after failed init

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:261]: Why not:
```suggestion
	if (sign < 0) {
```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:271]: MISRA - unwanted cast int -> char
```suggestion
		return '\0';
```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:763]:
```suggestion
			if ((flags & FLAG_PLUS) != 0) {
```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:738]:
```suggestion
		for (i = 0; i < s_len; i++) {
			feed(ctx, s[i]);
		}
```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:853]: Never use `l`! (or was it `1`? Hmm, I think it was `I`, can't tell ;))

File: stdio/bignum.c
Comment (agkaminski) [reply stdio/bignum.c:171]: My bad

File: stdio/bignum.c
Comment (agkaminski) [inline stdio/bignum.c:46]: Is there a reason for `(-1)` instead of plain `-1`? This differs from the style of the rest of the codebase.

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:97]:
```suggestion
			return -1;
```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:339]:
```suggestion
			*(carrier--) = '0';
```

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:381]: Comment about magic `13` would be nice

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:183]: `res` is never checked

File: stdio/format.c
Comment (agkaminski) [inline stdio/format.c:704]: MISRA rule 13.5 violation

Comment (agkaminski) [reply stdio/format.c:704]: And it even matters - potential `bd_backup` leak here

**Review by agkaminski — DISMISSED** [review]
LGTM, let's wait for a confirmation from @damianloew that the issues has been indeed been resolved

**Comment by astalke** [comment]
> Could you please check (or explain if you know) why when I tried to run the tests without commenting out the result was like:
> 
> ```
> (psh)% sysexec test-libc
> (psh)%
> ```
Quite likely this is a result of lack of memory.

> There are some issues found on `imxrt106x-evk` target:
> 
> ```c
> #include <stdio.h>
> 
> 
> int main(void)
> {
> 	printf("%.0f\n", 7.76639933231677154213e-02);
> 	printf("%.0e\n", 7.76639933231677154213e-02);
> 	printf("%.0g\n", 7.76639933231677154213e-02);
> 	printf("%.0a\n", 7.76639933231677154213e-02);
> 	printf("%.0f\n", 4.75);
> 	printf("%.0e\n", 4.75);
> 	printf("%.0g\n", 4.75);
> 	printf("%.0a\n", 4.75);
> 	printf("%.0f\n", 9.0);
> 	printf("%.0e\n", 9.0);
> 	printf("%.0g\n", 9.0);
> 	printf("%.0a\n", 9.0);
> 
> 	return 0;
> }
> ```
> 
> Screenshot from 2023-02-28 16-29-16
> 
> On `armv7a9-zynq7000-zedboard` works fine:
> 
> Screenshot from 2023-02-28 16-30-21
> 
> I suppose that probably the problem is in kernel (?), but leaving `request changes` until it's unresolved.
It's strange, I'll look into it

**Comment by astalke** [comment]
@damianloew I've just run it on hardware. It works fine on `armv7m7-imxrt106x-evk`

Zrzut ekranu z 2023-03-01 09-29-47

**Comment by astalke** [comment]
@damianloew I've updated [tests](https://github.com/phoenix-rtos/phoenix-rtos-project/pull/613).

**Comment by damianloew** [comment]
> @damianloew I've just run it on hardware. It works fine on `armv7m7-imxrt106x-evk`
> 
> Zrzut ekranu z 2023-03-01 09-29-47

Yeah, false alarm, the presented behavior was seen when using some older version of this branch, so sorry for that. Thanks for checking it, I see that you have already created PR in phoenix-rtos-project, but imxrt targets are excluded for test-libc. I would appreciate making this test passed for both imxrt106x and imxrt117x. We could also check stm32l4x6 and riscv targets, if everything is ok we can merge it - it can be easily done by using the new runner implementation (rebase onto jsarzy/runner), where those targets are added.

**Comment by niewim19** [comment]
Referencing connected issues:
 - https://github.com/phoenix-rtos/phoenix-rtos-project/issues/277

**Review by agkaminski — DISMISSED** [review]
LGTM

**Review by damianloew — APPROVED** [review]
Thanks for checking it on other targets by creating PR in phoenix-rtos-project. LGTM, but please rebase before a merge.

**Review by agkaminski — APPROVED** [review]

**Review by damianloew — APPROVED** [review]

**Pull request merged**
